### PR TITLE
Display bookmarks in collections sidebar

### PR DESCRIPTION
**Pull request by gusaiani**
In this PR we do display the bookmarks in a Collection's sidebar.

Also, because Bookmarks will now be displayed in that sidebar, we move and rename files so that they better reflect where they are.

Finally, we (I had done this…) had directories of non-container React components inside the the sidebar container directory. 
This way of organizing things did not gain any traction, and I agree it's not ideal — what we have is improvable in other ways. So here I am reverting to the customary directory organization, with non-container components in a different branch of the dir structure than the container.

### How to Test

1. Create a question.
2. Save the question.
3. Visit the question.
4. Open its left sidebar
5. Click on the ribbon icon, the rightmost icon in the list

image

It should toggle between black and blue as we click.
Blue means bookmarked.
Black means not bookmarked.

1. Without reloading the page, click on Metabase's icon on the main Nav.
2. Click on `Browse all items`

If you have bookmarked items, they should be rendered on the sidebar.

Refresh the page and you should still see the Bookmarks list.

### CODE REVIEW

File: frontend/src/metabase/collections/components/CollectionSidebar/Bookmarks/Bookmarks.jsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+      <BookmarkLinkRoot>
+        {bookmarks.map(({ id, name, type }, index) => {
+          const url = Urls.bookmark({ id, name, type });
+          return (
+            <Link
+              key={`bookmark-${id}`}
+              to={url}
+              selected={false}
+              onClick={() => {}}
```
Comment (ranquild) [inline frontend/src/metabase/collections/components/CollectionSidebar/Bookmarks/Bookmarks.jsx:54]: Is this by design?

File: frontend/src/metabase/collections/components/CollectionSidebar/Bookmarks/Bookmarks.jsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+              role="sidebar-bookmark"
```
Comment (ranquild) [inline frontend/src/metabase/collections/components/CollectionSidebar/Bookmarks/Bookmarks.jsx:55]: I'm not sure but based on MDN this ARIA role doesn't exist https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles.

File: frontend/src/metabase/query_builder/containers/QueryBuilder.jsx [hunk 3]
```diff
@@ -153,8 +153,8 @@ const mapStateToProps = (state, props) => {
 const mapDispatchToProps = {
   ...actions,
   onChangeLocation: push,
-  createBookmark,
-  deleteBookmark,
+  createBookmark: async id => Bookmark.actions.create({ id }),
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/containers/QueryBuilder.jsx:156]: There is no `await`, do we need `async` here?

File: frontend/src/metabase/lib/urls.js [hunk 4]
```diff
@@ -346,3 +346,22 @@ export function extractCollectionId(slug) {
...
+/*
+ * Will transform a name like `This name has spaces and Uppercases`
+ * into `this-name-has-spaced-and-uppercases`
+ *
+ * then prepend an entity type, say "card" or "dashboard"
+ * plus the passed id.
+ * "
+ */
+export function bookmark({ type, id, name }) {
+  const kebabCaseName = name
```
Comment (ranquild) [inline frontend/src/metabase/lib/urls.js:359]: Could `slugg` function be used here instead? Example usage https://github.com/metabase/metabase/blob/570e4a19e9ff0abcb45cc5c442f4e661cc58de5e/frontend/src/metabase/lib/urls.js#L118

File: frontend/src/metabase/collections/components/CollectionSidebar/Bookmarks/Bookmarks.jsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    <LabelContainer>
+      <BookmarkTypeIcon name="grid" />
+      {name}
+    </LabelContainer>
+  );
+};
+
+Label.propTypes = LabelPropTypes;
+
+const CollectionSidebarBookmarks = ({ bookmarks }) => {
```
Comment (ranquild) [inline frontend/src/metabase/collections/components/CollectionSidebar/Bookmarks/Bookmarks.jsx:37]: This is a new component, can we implement it using TypeScript? You can use `metabase-types/api` package to store global types, like we do here for timelines https://github.com/metabase/metabase/blob/eb9ef86f68d53374f7d621755ed802c43707027b/frontend/src/metabase-types/api/timeline.ts. Just create a file `bookmark.ts` with type definitions for bookmarks.

**Review by ranquild — APPROVED** [review]
Working as described. Left a couple of comments, please take a look 👍

**Comment by gusaiani** [comment]
Thank you @ranquild and @kdoh.

Alex, great points, appreciated. I'll implement them in a follow-up PR.

**Pull request merged**